### DISCUSSION TITLE
How can I use aws-sdk and sigv4 gems?

### DISCUSSION THREAD

**Discussion by WebDeverDan** [aws/aws-sdk-ruby#2757] [Q&A]
Greetings! I'm a junior dev and am working on a custom api to integrate with one of Amazon's systems. We've been told that we should use aws-sdk gem for grabbing credentials via our environment variables, and the sigv4 gem for creating our signature. Additionally we're using HTTParty for our api requests. I've successfully done this via the Postman AWS Signature auth headers, however am not finding any good examples of how to implement this in an actual application. As a junior dev, I'm also having extreme difficulty understanding the documentation and how to use the gems. Any help would be appreciated. Thanks!

**Comment by mullermp** [comment]
Thanks for opening a discussion. For SigV4 signing, you should install and use the aws-sigv4 gem and use `sign_request` method - https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/Sigv4/Signer.html . As far as credentials loading goes, it depends on what type of credentials you are using. If you're using static credentials (i.e.access key id, secret access key, etc) you can fetch the values from ENV yourself and pass them to the sigv4 signer. If you have more complex needs, like credentials on EC2, you could use aws-sdk-core's credential providers that will automatically fetch these, and you can pass those providers to the signer as well.

**Reply by WebDeverDan** [reply]
Thanks for responding. Would you have an example of how to use the sign_request? I'm not exactly sure how/where I should use that within the get requests I'm trying to do. 

For the credentials, we're using assume role credentials in our ENV (for testing I generated via the AWS SDK CLI assume role functionality, resulting in expiring access_key, secret_access_key, and secret_access_token). Thanks so much!

**Reply by mullermp** [reply]
Here's an example of how we sign SDK requests: https://github.com/aws/aws-sdk-ruby/blob/aaefb3f04925921189b1df4a7e203c8290fadddf/gems/aws-sdk-core/lib/aws-sdk-core/plugins/signature_v4.rb#L121-L126

We pass it stuff from the HTTP request (your HTTParty request object, or Net::HTTP request object, should have a method, url, headers, body). Then we apply the signature to the headers.

For credentials, you can use AssumeRoleCredentials from aws-sdk-core to easily fetch/manage assumed role credentials and pass it to the sigv4 signer - https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb

**Reply by WebDeverDan** [reply]
Thanks so much! I'll work through this and come back if i still need additional assistance.

**Reply by mullermp** [reply]
You're getting closer. Here are some things to keep in mind:
* You should only need to pass headers to `sign_request` that you actually want signed. You shouldn't need to pass things like content-type and your API key unless it's required in signing, but you should send them in your actual request.
* You need to apply the signature. When you call sign_request, you'll get a Signature object that has: host, x-amz-date, x-amz-security-token, x-amz-content-sha256, and authorization headers. All 5 headers should be added to your http request (you should be able to do http_request.headers.merge(signature.headers) or equivalent).

**Reply by WebDeverDan** [reply]
So I was able to access the header information, although now am getting a 403. RESPONSE: 403 | {"message":"The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method.

Can you see anything in here that would cause this? Thank you!

role_credentials = Aws::AssumeRoleCredentials.new(
client: Aws::STS::Client.new,
# role arn will be hidden to env
role_arn: "XXX",
role_session_name: "default"
)
signer = Aws::Sigv4::Signer.new(
service: 'execute-api',
region: 'us-east-1',
credentials_provider: role_credentials
)
signature = signer.sign_request(
http_method: 'GET',
url: "#{V1::API::HOST_URL}/#{endpoint}"
)
@headers = DEFAULT_HEADERS.merge!(signature.headers)
